### PR TITLE
sse: implement the initial version

### PR DESCRIPTION
**Pull request by wuhan005**
https://github.com/flamego/flamego/issues/85

### CODE REVIEW

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        uses: golangci/golangci-lint-action@v2
+        with:
+          version: latest
+          args: --timeout=30m
+
+  test:
+    name: Test
+    strategy:
+      matrix:
+        go-version: [ 1.16.x, 1.17.x ]
```
Comment (unknwon) [inline .github/workflows/go.yml:38]:
```suggestion
        go-version: [ 1.16.x, 1.17.x, 1.18.x ]
```

File: codecov.yml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+coverage:
+  range: "60...95"
+  status:
+    project:
+      default:
+        threshold: 1%
+
+comment:
+  layout: 'diff'
```
Comment (unknwon) [inline codecov.yml:9]:
```suggestion
  layout: 'diff'

github_checks: false
```

Let's stop letting codecov fail CI.

File: sse.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+const (
+	defaultPingInterval = 10 * time.Second
+)
+
+type Options struct {
+	// PingInterval is the time to wait between sending pings to the client.
+	PingInterval time.Duration
+}
+
+type Connection struct {
```
Comment (unknwon) [inline sse.go:25]: Does this type need to be public? If so we need docstring, if not let's make it private.

File: sse.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	"time"
+
+	"github.com/flamego/flamego"
+)
+
+const (
+	defaultPingInterval = 10 * time.Second
+)
+
+type Options struct {
```
Comment (unknwon) [inline sse.go:20]: Missing docstring.

File: sse.go [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+	"net/http"
+	"reflect"
+	"time"
+
+	"github.com/flamego/flamego"
+)
+
+const (
+	defaultPingInterval = 10 * time.Second
+)
```
Comment (unknwon) [inline sse.go:18]: If this is private _and_ only used once, I don't see a reason to make it a package-level constant.

File: sse.go [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+type Options struct {
+	// PingInterval is the time to wait between sending pings to the client.
+	PingInterval time.Duration
```
Comment (unknwon) [inline sse.go:22]:
```suggestion
	// PingInterval is the time to wait between sending pings to the client.
	// Default is 10 seconds.
	PingInterval time.Duration
```

File: sse.go [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+	Sender reflect.Value
+
+	// context is the request context.
+	context flamego.Context
+
+	// ticker for pinging the client.
+	ticker *time.Ticker
+}
+
+func Handler(bind interface{}, options ...*Options) flamego.Handler {
```
Comment (unknwon) [inline sse.go:40]: Why do we need pointer to the `Options`? Doesn't seem like being consistent with other middleware, unless we have a very good reason to do so.

File: sse.go [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+	}
+
+	c.write("event: error\ndata: eof\n\n")
+	c.flush()
+	c.write("events: stream closed")
+	c.flush()
+}
+
+// makeChanOfType create a chan of the given type as a reflect.Value.
+func makeChanOfType(typ reflect.Type, chanBuffer int) reflect.Value {
```
Comment (unknwon) [inline sse.go:142]: This function seems only being called once, why it deserves a dedicated function but `reflect.ChanOf(reflect.SendDir, sse.Sender.Type().Elem())` doesn't (Line 52)?

File: sse.go [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+	c.write(": ping\n\n")
+	c.write("events: stream opened\n\n")
+	c.flush()
+
+	cases := make([]reflect.SelectCase, 3)
+	cases[senderSend] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: c.Sender, Send: reflect.ValueOf(nil)}
+	cases[tickerTick] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(c.ticker.C), Send: reflect.ValueOf(nil)}
+	cases[timeout] = reflect.SelectCase{Dir: reflect.SelectRecv, Chan: reflect.ValueOf(time.After(time.Hour)), Send: reflect.ValueOf(nil)}
+
+L:
```
Comment (unknwon) [inline sse.go:107]:
```suggestion
loop:
```

File: sse.go [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+	for {
+		chosen, message, ok := reflect.Select(cases)
+		switch chosen {
+
```
Comment (unknwon) [inline sse.go:111]:
```suggestion
```

File: sse.go [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+		case senderSend:
+			if !ok {
+				// Sender channel has been closed.
+				return
+			}
+
+			c.write("data: ")
+			evt, _ := json.Marshal(message.Interface())
```
Comment (unknwon) [inline sse.go:119]: What's the worst the thing that could happen when we ignore the error here?

**Review by unknwon — CHANGES_REQUESTED** [review]
Thanks!

Can we get some tests as well?

File: sse.go (hunk 3 above)
Comment (wuhan005) [reply sse.go:26]: Hmmm, it should be private I think.

**Review by unknwon — APPROVED** [review]

**Pull request merged**